app/vmselect/netstorage/netstorage.go

```diff
@@ -1441,6 +1527,15 @@ func (tbfw *tmpBlocksFileWrapper) RegisterAndWriteBlock(mb *storage.MetricBlock,
 	m := tbfwLocal.m
 	metricName := mb.MetricName
 	addrsIdx := tbfwLocal.prevAddrsIdx
+
+	if isMultiTenant {
+		accountID := encoding.UnmarshalUint32(metricName[:4])
+		projectID := encoding.UnmarshalUint32(metricName[4:8])
```

Comment:
This code is unsafe and fragile. It uses internal knowledge of storage.Metric marshal implementation. 

It's better to mutate metricName during results processing at RunParallel.

 `shouldConvertTenantToLabels` field could be added to the `Results` struct.

And be checked before calling callback at RunParallel.